### PR TITLE
Fix/bulk loader nullstring

### LINKED ISSUE #1938: PostgreSQL Bulkload writes caveat names as null instead of empty string
### What platforms are affected?

linux, macos, windows, others

### What architectures are affected?

amd64, arm64, others

### What SpiceDB version are you using?

v.1.33.0

### Steps to Reproduce

1. Create Backup with CLI Tool (`zed backup create backup.bin`)
2. Restore Backup with CLI Tool (`zed backup restore backup.bin`)

But this will be the same over the normal bulkimport endpoint aswell.

### Expected Result

Writen Relationtuples in the database should be the same as in the source.

### Actual Result

The bulk imported tuples without a caveat have the `caveatname` as `null` value in the database. But this should be an empty string. This is problematic when we try to update an existing relation with a caveat later on because the sql check (`caveat_name <> 'some_caveat_name'`) fails for a null value.

It looks like the BulkLoad function uses a sql.NullString instead of a simple string, this causes the problem.

### PR DESCRIPTION
**Pull request by heissa83**
Fixes #1938 and also fixes problems like it in the future by using `IS DISTINCT FROM`instead of `<>` to compare a potential null value field. (This will make it so that already bulk loaded relations will work correctly).

### CODE REVIEW

**Comment by heissa83** [comment]
recheck

**Comment by heissa83** [comment]
I added a new test that checks that caveated relations can be updated after they were added by a BulkLoad. I also did the negative test locally and both changes i made fix the problem. (The test also showed that i was using the wrong column name in the first place, i also fixed that)

**Review by vroldanbet — CHANGES_REQUESTED** [review]
one of the tests is failing

File: pkg/datastore/test/bulk.go [hunk 1]
```diff
@@ -119,6 +119,85 @@ func BulkUploadAlreadyExistsSameCallErrorTest(t *testing.T, tester DatastoreTest
...
+	_, err = ds.ReadWriteTx(ctx, func(ctx context.Context, rwt datastore.ReadWriteTransaction) error {
+		loaded, err := rwt.BulkLoad(ctx, bulkSource)
+		require.NoError(err)
+		require.Equal(uint64(tc), loaded)
+		return err
+	})
+	require.NoError(err)
+
+	head, err := ds.HeadRevision(ctx)
+	require.NoError(err)
```
[the remark this answers is not in the document]
Comment (heissa83) [reply pkg/datastore/test/bulk.go]: Changed it to use the revision from the last write beforehand.

**Comment by heissa83** [comment]
Fixed the tests.
Sorry for some reason my docker daemon keeps crashing when i run the complete test suite locally.

File: internal/datastore/postgres/readwrite.go [hunk 2]
```diff
@@ -725,7 +725,7 @@ func exactRelationshipDifferentCaveatClause(r *core.RelationTuple) sq.And {
 			colUsersetRelation:  r.Subject.Relation,
 		},
 		sq.Or{
-			sq.Expr(`? IS DISTINCT FROM ?`, colCaveatContextName, caveatName),
```
Comment (vroldanbet) [inline internal/datastore/postgres/readwrite.go]: Can you describe why this fixes the test issue? Both look equivalent to me

Comment (heissa83) [reply internal/datastore/postgres/readwrite.go]: > Can you describe why this fixes the test issue? Both look equivalent to me


In the first one (`? IS DISTINCT FROM ?`) the `caveat_name` parameter is interpreted as a string value, and so all the caveat names get checked against that, and they are always distinct from that (if you dont try to use a caveat named `caveat_name`). If it is directly written into the Expression (`caveat_name IS DISTINCT FROM ?`) it gets interpreted as the column name and the value of that is used to make the comparison.

So the difference is one will look like this in sql `"caveat_name" IS DISTINCT FROM "some_caveat_name"` and the other looks like `caveat_name IS DISTINCT FROM "some_caveat_name"` in the end.

**Review by josephschorr — DISMISSED** [review]
LGTM

File: internal/datastore/postgres/readwrite.go (hunk 2 above)
Comment (vroldanbet) [reply internal/datastore/postgres/readwrite.go]: makes sense, thanks for clarifying!

**Comment by heissa83** [comment]
@vroldanbet do you still have a change request? If not can this PR be merged as it is right now?

**Review by vroldanbet — APPROVED** [review]
LGTM!

**Pull request merged**